File: TwoPointer/19.remove-nth-node-from-end-of-list.go
package leetcode

// @leet start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	dummy := &ListNode{Next: head}
	slow, fast := dummy, dummy
	for range n {
		fast = fast.Next
	}
	for fast != nil {
		fast = fast.Next
		slow = slow.Next
	}
	slow.Next = slow.Next.Next

	return dummy.Next
}

// @leet end

File: TwoPointer/19.remove-nth-node-from-end-of-list.py
from typing import Optional, cast
from type import ListNode


# @leet start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        # fast = slow = dummy.next # NOTE: dummy head effetc
        fast = slow = dummy
        for _ in range(n):
            fast = fast.next

        while fast.Next is not None:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next

        return dummy.next


# @leet end

